pkg/apis/serving/v1alpha1/service_types.go

```diff
@@ -0,0 +1,156 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ServiceSpec defines the desired state of Service
+type ServiceSpec struct {
+	// If specified, passes name to generated Revision.
+	RevisionName string `json:"rollout,omitempty"`
+	// Must contain either one or two elements.
+	// The first is the "active" revisions and the second is the "candidate" revision.
+	Revisions []string `json:"revisions,omitempty"`
+	// May only be specified if revisions has two elements. Defaults to 0.
+	// This percentage of traffic is routed to the second revision, the remainder to the first.
+	RolloutPercent int `json:"rolloutPercent,omitempty"`
+
+	MinReplicas string `json:"minReplicas,omitempty"`
+	MaxReplicas string `json:"maxReplicas,omitempty"`
+
+	/*
+		The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	*/
+	Custom      *CustomSpec      `json:"custom,omitempty"`
+	Tensorflow  *TensorflowSpec  `json:"tensorflow,omitempty"`
+	XGBoost     *XGBoostSpec     `json:"xgBoost,omitempty"`
+	ScikitLearn *ScikitLearnSpec `json:"scikitLearn,omitempty"`
+}
+
+// TensorflowSpec defines arguments for configuring Tensorflow model serving.
+type TensorflowSpec struct {
+	ModelUri string `json:"modelUri"`
+	// Defaults to latest TF Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// XGBoostSpec defines arguments for configuring XGBoost model serving.
+type XGBoostSpec struct {
+	ModelUri string `json:"modelUri"`
+	// Defaults to latest XGBoost Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	//Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// ScikitLearnSpec defines arguments for configuring ScikitLearn model serving.
+type ScikitLearnSpec struct {
+	ModelURI string `json:"modelUri"`
+	// Defaults to latest ScikitLearn Version.
+	RuntimeVersion string `json:"runtimeVersion,omitempty"`
+	// Defaults to requests and limits of 1CPU, 2Gb MEM.
+	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// CustomSpec provides a hook for arbitrary container configuration.
+type CustomSpec struct {
+	Container v1.Container `json:"container"`
+	// Defaults to /healthz.
+	LivenessProbe *v1.Probe `json:"livenessProbe,omitempty"`
```

Comment:
You're right. I looked at the rest of PodSpec. The only other field we might want to consider is serviceAccount.

I was thinking about the Container expose question. I decided to add it to prompt discussion. What do you think?

Reply:
Exposing `Container` opens the door to run arbitray image, so what we did is to leverage knative build which defines a build template to turn the prediction function(e.g https://github.com/yuzisun/helloworld/blob/master/test/helloworld.py) to a serving image and orchestrate the source to image job on kubernetes. What do you guys think ?

This is an example of our custom spec
``` 
    dockerSecret: docker-credential
    image:  test/knative-iaas:latest 
    runtime: python-3
    source:
      git:
        gitSecret: git-credential
        url: https://github.com/yuzisun/helloworld
        revision: master
    readinessProbe 
    livenessProbe
    resources
    envs
```

Reply:
I see arbitrary image as desirable. I'm completely open to having both (as per Knative). What if we had 

`custom.container`
XOR
`custom.source`

Reply:
I think both would be good to allow people to use existing Images they build outside KNative. This would also allow people to use Images build using Seldon to be incorporated hopefully.

Reply:
Agree with @cliveseldon I think we should allow users not to use KNative.

Reply:
@gaocegege, regarding Knative Configuration spec, this was discussed in the original proposal, but we agreed against this in the MLSpec PR.

I'm happy with either of the following
* K8s Pod Spec
* K8s Container Spec

As well as another option for
* SourceToUrl Spec (We should discuss this in another PR).

I may remove this so we can make parallel progress on the CRD. What do you think?

Reply:
SGTM, thanks